File: frontend/lib/domain/models/book.dart
/// Domain models for the book reader.

class BookData {
  final int id;
  final String title;
  final int totalPages;
  final int manifestVersion;
  final bool isPublished;

  BookData({
    required this.id,
    required this.title,
    required this.totalPages,
    required this.manifestVersion,
    required this.isPublished,
  });

  factory BookData.fromJson(Map<String, dynamic> json) => BookData(
    id: json['id'] ?? 0,
    title: json['title'] ?? '',
    totalPages: json['total_pages'] ?? 0,
    manifestVersion: json['manifest_version'] ?? 0,
    isPublished: json['is_published'] ?? false,
  );
}

class Chapter {
  final int id;
  final String title;
  final String? titleAr;
  final int sortOrder;
  final int? startPage;
  final int? endPage;

  Chapter({
    required this.id,
    required this.title,
    this.titleAr,
    required this.sortOrder,
    this.startPage,
    this.endPage,
  });

  factory Chapter.fromJson(Map<String, dynamic> json) => Chapter(
    id: json['id'] ?? 0,
    title: json['title'] ?? '',
    titleAr: json['title_ar'],
    sortOrder: json['sort_order'] ?? 0,
    startPage: json['start_page'],
    endPage: json['end_page'],
  );
}

class PageSummary {
  final int id;
  final int pageNumber;
  final String? imageUrl;
  final bool hasTextData;
  final bool isAnnotated;

  PageSummary({
    required this.id,
    required this.pageNumber,
    this.imageUrl,
    required this.hasTextData,
    required this.isAnnotated,
  });

  factory PageSummary.fromJson(Map<String, dynamic> json) => PageSummary(
    id: json['id'] ?? 0,
    pageNumber: json['page_number'] ?? 0,
    imageUrl: json['image_url'],
    hasTextData: json['has_text_data'] ?? false,
    isAnnotated: json['is_annotated'] ?? false,
  );
}

class PageDetail {
  final int id;
  final int pageNumber;
  final String layoutType; // "pdf" | "native"
  final String? imageUrl;
  final String? image2xUrl;
  final String? sourceImageUrl;
  final int? imageWidth;
  final int? imageHeight;
  final bool hasTextData;
  final bool isAnnotated;
  final String analysisStatus; // empty|pending|analyzing|draft|published|error
  final List<TextUnit> textUnits;
  final List<Section> sections;
  final String? audioUrl;
  final List<String> audioUrls;

  PageDetail({
    required this.id,
    required this.pageNumber,
    this.layoutType = 'pdf',
    this.imageUrl,
    this.image2xUrl,
    this.sourceImageUrl,
    this.imageWidth,
    this.imageHeight,
    required this.hasTextData,
    required this.isAnnotated,
    this.analysisStatus = 'empty',
    required this.textUnits,
    this.sections = const [],
    this.audioUrl,
    this.audioUrls = const [],
  });

  /// True when page has an uploaded image and bbox-positioned units
  bool get hasOverlayData =>
      sourceImageUrl != null &&
      sourceImageUrl!.isNotEmpty &&
      textUnits.any((u) => u.bboxW > 0);

  factory PageDetail.fromJson(Map<String, dynamic> json) => PageDetail(
    id: json['id'] ?? 0,
    pageNumber: json['page_number'] ?? 0,
    layoutType: json['layout_type'] ?? 'pdf',
    imageUrl: json['image_url'],
    image2xUrl: json['image_2x_url'],
    sourceImageUrl: json['source_image_url'],
    imageWidth: json['image_width'],
    imageHeight: json['image_height'],
    hasTextData: json['has_text_data'] ?? false,
    isAnnotated: json['is_annotated'] ?? false,
    analysisStatus: json['analysis_status'] ?? 'empty',
    textUnits: (json['text_units'] as List?)
        ?.map((e) => TextUnit.fromJson(e))
        .toList() ?? [],
    sections: (json['sections'] as List?)
        ?.map((e) => Section.fromJson(e))
        .toList() ?? [],
    audioUrl: json['audio_url'],
    audioUrls: (json['audio_urls'] as List?)
        ?.map((e) => e as String)
        .toList() ?? [],
  );
}

class TextUnit {
  final int id;
  final String unitType;
  final String textContent;
  final double bboxX;
  final double bboxY;
  final double bboxW;
  final double bboxH;
  final int sortOrder;
  final bool isManual;
  final double? confidence;
  final String? audioSegmentUrl;
  final Map<String, dynamic> metadata;

  TextUnit({
    required this.id,
    required this.unitType,
    required this.textContent,
    required this.bboxX,
    required this.bboxY,
    required this.bboxW,
    required this.bboxH,
    required this.sortOrder,
    required this.isManual,
    this.confidence,
    this.audioSegmentUrl,
    this.metadata = const {},
  });

  factory TextUnit.fromJson(Map<String, dynamic> json) => TextUnit(
    id: json['id'] ?? 0,
    unitType: json['unit_type'] ?? 'letter',
    textContent: json['text_content'] ?? '',
    bboxX: (json['bbox_x'] ?? 0).toDouble(),
    bboxY: (json['bbox_y'] ?? 0).toDouble(),
    bboxW: (json['bbox_w'] ?? 0).toDouble(),
    bboxH: (json['bbox_h'] ?? 0).toDouble(),
    sortOrder: json['sort_order'] ?? 0,
    isManual: json['is_manual'] ?? false,
    confidence: (json['confidence'] as num?)?.toDouble(),
    audioSegmentUrl: json['audio_segment_url'],
    metadata: (json['metadata'] as Map<String, dynamic>?) ?? {},
  );

  /// Get section from metadata
  String get section => (metadata['section'] as String?) ?? '';

  /// Get label from metadata
  String get label => (metadata['label'] as String?) ?? textContent;

  /// Get grid position from metadata
  Map<String, dynamic>? get grid => metadata['grid'] as Map<String, dynamic>?;

  /// Get render info from metadata
  Map<String, dynamic>? get renderInfo => metadata['render'] as Map<String, dynamic>?;
}


/// A section groups related text units into a pedagogical block.
class Section {
  final int id;
  final String sectionType;
  final String? targetLetter;
  final String? titleAr;
  final String? titleUz;
  final int sortOrder;
  final List<int> unitIds;
  final double? bboxYStart;
  final double? bboxYEnd;
  final bool isManual;

  Section({
    required this.id,
    required this.sectionType,
    this.targetLetter,
    this.titleAr,
    this.titleUz,
    required this.sortOrder,
    required this.unitIds,
    this.bboxYStart,
    this.bboxYEnd,
    this.isManual = false,
  });

  factory Section.fromJson(Map<String, dynamic> json) => Section(
    id: json['id'] ?? 0,
    sectionType: json['section_type'] ?? 'generic',
    targetLetter: json['target_letter'],
    titleAr: json['title_ar'],
    titleUz: json['title_uz'],
    sortOrder: json['sort_order'] ?? 0,
    unitIds: (json['unit_ids'] as List?)?.map((e) => e as int).toList() ?? [],
    bboxYStart: (json['bbox_y_start'] as num?)?.toDouble(),
    bboxYEnd: (json['bbox_y_end'] as num?)?.toDouble(),
    isManual: json['is_manual'] ?? false,
  );

  /// Display title — prefer Uzbek, fallback to Arabic
  String get displayTitle {
    if (titleUz != null && titleUz!.isNotEmpty) return titleUz!;
    if (titleAr != null && titleAr!.isNotEmpty) return titleAr!;
    return sectionType;
  }

  /// Short chip label for navigator
  String get chipLabel {
    if (targetLetter != null && targetLetter!.isNotEmpty) return targetLetter!;
    return displayTitle;
  }
}

File: backend/app/api/v1/book.py
"""Public book API endpoints."""

from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.section import Section

from app.database import get_db
from app.models.book import Book, Chapter, Page, TextUnit
from app.models.audio import AudioFile, UnitSegmentMapping, AudioSegment
from app.config import get_settings
from app.schemas import BookOut, BookSummary, PageOut, TextUnitOut, ChapterOut
from app.api.deps import get_published_book

router = APIRouter(prefix="/book", tags=["Book"])
settings = get_settings()


@router.get("", response_model=BookSummary)
async def get_book(book: Book = Depends(get_published_book)):
    """Get the book summary (single book app)."""
    return book


@router.get("/chapters", response_model=List[ChapterOut])
async def get_chapters(
    book: Book = Depends(get_published_book),
    db: AsyncSession = Depends(get_db),
):
    """Get table of contents (chapters)."""
    result = await db.execute(
        select(Chapter)
        .where(Chapter.book_id == book.id)
        .order_by(Chapter.sort_order)
    )
    return result.scalars().all()


@router.get("/pages", response_model=List[dict])
async def get_pages_list(
    book: Book = Depends(get_published_book),
    db: AsyncSession = Depends(get_db),
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """Get paginated list of pages (summaries)."""
    result = await db.execute(
        select(Page)
        .where(Page.book_id == book.id)
        .order_by(Page.page_number)
        .offset(offset)
        .limit(limit)
    )
    pages = result.scalars().all()

    return [
        {
            "id": p.id,
            "page_number": p.page_number,
            "image_url": f"{settings.MEDIA_BASE_URL}/{p.image_path}" if p.image_path else None,
            "has_text_data": p.has_text_data,
            "is_annotated": p.is_annotated,
        }
        for p in pages
    ]


@router.get("/pages/{page_number}", response_model=dict)
async def get_page(
    page_number: int,
    book: Book = Depends(get_published_book),
    db: AsyncSession = Depends(get_db),
):
    """Get a single page with all text units and their audio mappings."""

    result = await db.execute(
        select(Page)
        .options(
            selectinload(Page.text_units),
            selectinload(Page.sections),
        )
        .where(Page.book_id == book.id, Page.page_number == page_number)
    )
    page = result.scalar_one_or_none()
    if not page:
        raise HTTPException(status_code=404, detail="Sahifa topilmadi")

    # Sahifaga tegishli audio fayllarni topish
    audio_result = await db.execute(
        select(AudioFile)
        .where(
            AudioFile.book_id == book.id,
            AudioFile.page_start <= page_number,
            AudioFile.page_end >= page_number,
            AudioFile.status == "ready",
        )
        .order_by(AudioFile.id)
    )
    audio_files = audio_result.scalars().all()
    # Sahifaga tegishli audio URL'larning ro'yxati
    audio_urls = [
        f"{settings.MEDIA_BASE_URL}/{af.file_path}"
        for af in audio_files
        if af.file_path
    ]
    # Birinchi audio URL (asosiy)
    page_audio_url = audio_urls[0] if audio_urls else None

    # Build text units with audio URLs
    units = []
    for unit in sorted(page.text_units, key=lambda u: u.sort_order):
        # Get published audio mapping
        mapping_result = await db.execute(
            select(UnitSegmentMapping)
            .options(selectinload(UnitSegmentMapping.audio_segment))
            .where(
                UnitSegmentMapping.text_unit_id == unit.id,
                UnitSegmentMapping.is_published == True,
            )
            .limit(1)
        )
        mapping = mapping_result.scalar_one_or_none()
        audio_url = None
        if mapping and mapping.audio_segment and mapping.audio_segment.file_path:
            audio_url = f"{settings.MEDIA_BASE_URL}/{mapping.audio_segment.file_path}"

        units.append({
            "id": unit.id,
            "unit_type": unit.unit_type.value if hasattr(unit.unit_type, 'value') else unit.unit_type,
            "text_content": unit.text_content,
            "bbox_x": unit.bbox_x,
            "bbox_y": unit.bbox_y,
            "bbox_w": unit.bbox_w,
            "bbox_h": unit.bbox_h,
            "sort_order": unit.sort_order,
            "is_manual": unit.is_manual,
            "audio_segment_url": audio_url,
            "metadata": unit.metadata_ or {},
        })

    # Build sections
    sections = []
    for sec in sorted(page.sections, key=lambda s: s.sort_order):
        sections.append({
            "id": sec.id,
            "section_type": sec.section_type.value if hasattr(sec.section_type, 'value') else sec.section_type,
            "target_letter": sec.target_letter,
            "title_ar": sec.title_ar,
            "title_uz": sec.title_uz,
            "sort_order": sec.sort_order,
            "unit_ids": sec.unit_ids or [],
            "bbox_y_start": sec.bbox_y_start,
            "bbox_y_end": sec.bbox_y_end,
            "is_manual": sec.is_manual,
        })

    return {
        "id": page.id,
        "page_number": page.page_number,
        "layout_type": getattr(page, 'layout_type', 'pdf') or 'pdf',
        "image_url": f"{settings.MEDIA_BASE_URL}/{page.image_path}" if page.image_path else None,
        "image_2x_url": f"{settings.MEDIA_BASE_URL}/{page.image_2x_path}" if page.image_2x_path else None,
        "image_width": page.image_width,
        "image_height": page.image_height,
        "has_text_data": page.has_text_data,
        "is_annotated": page.is_annotated,
        "text_units": units,
        "sections": sections,
        "audio_url": page_audio_url,
        "audio_urls": audio_urls,
    }
